src/combinatorics/simple_graph/adj_matrix.lean

```diff
@@ -15,89 +16,257 @@ properties to computational properties of the matrix.
+namespace matrix
+
+/-- `A : matrix V V α` is qualified as an "adjacency matrix" if
+    (1) every entry of `A` is `0` or `1`,
+    (2) `A` is symmetric,
+    (3) every diagonal entry of `A` is `0`. -/
+structure adj_matrix [has_zero α] [has_one α] (A : matrix V V α) : Prop :=
+(zero_or_one : ∀ i j, (A i j) = 0 ∨ (A i j) = 1 . obviously)
+(symm : A.is_symm . obviously)
+(loopless : ∀ i, A i i = 0 . obviously)
+
+namespace adj_matrix
+
+lemma apply_diag [has_zero α] [has_one α]
+{A : matrix V V α} (h : adj_matrix A) (i : V) :
+  A i i = 0 :=
+by simp [h.loopless]
+
+@[simp]
+lemma apply_diag_ne [mul_zero_one_class α] [nontrivial α]
+{A : matrix V V α} (h : adj_matrix A) (i : V) :
+  ¬ A i i = 1 :=
+by simp [h.apply_diag i]
+
+@[simp]
+lemma apply_ne_one_iff [mul_zero_one_class α] [nontrivial α]
+{A : matrix V V α} (h : adj_matrix A) (i j : V) :
+  ¬ A i j = 1 ↔ A i j = 0 :=
+begin
+  replace h := h.zero_or_one i j,
+  split,
+  { tauto },
+  { rintros g, simp [g] }
+end
+
+@[simp]
+lemma apply_ne_zero_iff [mul_zero_one_class α] [nontrivial α]
+{A : matrix V V α} (h : adj_matrix A) (i j : V) :
+  ¬ A i j = 0 ↔ A i j = 1 :=
+begin
+  replace h := h.zero_or_one i j,
+  split,
+  { tauto },
+  { rintros g, simp [g] }
+end
+
+/-- For `A : matrix V V α` and `h : adj_matrix A`,
+    `h.to_graph` is the simple graph induced by `A`. -/
+def to_graph
+[mul_zero_one_class α] [nontrivial α] [decidable_eq α]
+{A : matrix V V α} (h : adj_matrix A) :
+  simple_graph V :=
+{ adj := λ i j, ite (A i j = 1) true false,
+  sym := λ i j hij, by simp only [h.symm.apply i j]; convert hij,
+  loopless := λ i, by simp [h] }
+
+instance [mul_zero_one_class α] [nontrivial α] [decidable_eq α]
+{A : matrix V V α} (h : adj_matrix A) :
+  decidable_rel h.to_graph.adj :=
+by {simp [to_graph], apply_instance}
+
+end adj_matrix
+
+/-- For `A : matrix V V α`, `A.compl` is supposed to be the adjacency matrix of
+    the complement graph of the graph induced by `A.adj_matrix`. -/
+def compl [has_zero α] [has_one α] [decidable_eq α] [decidable_eq V] (A : matrix V V α)  :
+  matrix V V α :=
+λ i j, ite (i = j) 0 (ite (A i j = 0) 1 0)
+
+@[simp]
+lemma compl_apply_diag [has_zero α] [has_one α] [decidable_eq α] [decidable_eq V]
+(A : matrix V V α) (i : V) :
+  A.compl i i = 0 :=
+by simp [compl]
+
+@[simp]
+lemma compl_apply [has_zero α] [has_one α] [decidable_eq α] [decidable_eq V]
+(A : matrix V V α) (i j : V) :
+  A.compl i j = 0 ∨ A.compl i j = 1 :=
+begin
+  unfold compl,
+  by_cases h : i = j, {simp [h]},
+  by_cases g : A i j = 0; simp [h , g]
+end
+
+@[simp]
+lemma symm_compl [has_zero α] [has_one α] [decidable_eq α] [decidable_eq V]
+(A : matrix V V α) (h : A.is_symm) :
+  A.compl.is_symm :=
+begin
+  ext,
+  unfold compl,
+  congr' 1, {tidy},
+  congr' 1,
+  apply propext,
+  rw [h.apply i j]
+end
+
+@[simp]
+lemma adj_matrix_compl [has_zero α] [has_one α] [decidable_eq α] [decidable_eq V]
+(A : matrix V V α) (h : A.is_symm) :
+  adj_matrix A.compl :=
+{ symm := by simp [h] }
```

Comment:
You can move `compl_apply_diag`, `compl_apply`, and `symm_compl` into this lemma and delete them, since they're all stating one of the axioms of an adjacency matrix.

Reply:
Is it better to combine them as one than writing them separately?

Reply:
Never mind, I thought `@[simps]` would work for this lemma, but it doesn't, so it's better to write them out separately.  Could you change `@[simps]` for `is_adj_matrix_adj_matrix` and `is_adj_matrix_compl` back to `@[simp]`?